### PR TITLE
[ISSUE-112] Add total_shards_per_node setting to allocate in ILM

### LINKED ISSUE #112: [Bug] `total_shards_per_node` isn't supported by the `allocate` action of `elasticstack_elasticsearch_index_lifecycle`
**Describe the bug**
According to the [ILM docs](https://www.elastic.co/guide/en/elasticsearch/reference/7.17/ilm-allocate.html), the `allocate` action supports the `total_shards_per_node` param. This TF provider, however, doesn't which produces an error when used.

**To Reproduce**
1. TF configuration used:
```
resource "elasticstack_elasticsearch_index_lifecycle" "ilm_policy" {

  name = "ilm_policy"

  hot {
    min_age = "0m"
  }

  warm {
    min_age = "27h"
    allocate {
      include = jsonencode({
        _tier_preference = "data_cold,data_warm,data_hot"
      })
      require = jsonencode({
        es_lifecycle_role = "cold"
      })
      total_shards_per_node = 2
    }
    forcemerge {
      max_num_segments = 1
    }
  }

  cold {
    min_age = "3d"
    allocate {
      include = jsonencode({
        _tier_preference = "data_cold,data_warm,data_hot"
      })
      require = jsonencode({
        es_lifecycle_role = "cold"
      })
    }
  }
}
```
3. Execute `terraform plan`
4. See the error in the output:
```
╷
│ Error: Unsupported argument
│
│   on ilm_policy.tf line 18, in resource "elasticstack_elasticsearch_index_lifecycle" "ilm_policy":
│   18:       total_shards_per_node = 2
│
│ An argument named "total_shards_per_node" is not expected here.
╵
Operation failed: failed running terraform plan (exit 1)
```

**Expected behavior**
`elasticstack_elasticsearch_index_lifecycle` resource is expected to support the `total_shards_per_node` param as part of its `allocate` action.

**Versions:**
 - OS: Linux
 - Terraform Version 1.1.7
 - Provider version v0.3.3
 - Elasticsearch Version 7.17.3

### PR DESCRIPTION
**Pull request by RobsonSutton**
- Fix small typo on variable name (housekeeping)
- Add total_shards_per_node setting (identified as missing: https://github.com/elastic/terraform-provider-elasticstack/issues/112) 

Fixes #112

### CODE REVIEW

**Comment by RobsonSutton** [comment]
@Kushmaro / @olksdr - Hi both 👋 Please can you review and provide some feedback on this one when you get a chance?

P.S. New to Go so apologies if anything has been missed...!

**Comment by tobio** [comment]
jenkins test this

**Comment by tobio** [comment]
jenkins test this

**Comment by tobio** [comment]
@RobsonSutton the CI test suite is run against a range of ES versions (you can see the pipeline definition [here](https://github.com/elastic/terraform-provider-elasticstack/blob/main/.ci/pipelines/acceptance.groovy#L14)). 

I think we'll want to only define this property in the acceptance tests when they're being run against a supported ES version.

**Comment by RobsonSutton** [comment]
@tobio - Apologies for the delay! Yeah I didn't have time the other day to figure this one out but hopefully I should get some time later today to try figure out the best way to skip these for specific versions within the tests 👍

**Comment by tobio** [comment]
jenkins test this

**Comment by tobio** [comment]
jenkins test this

**Comment by othmane399** [comment]
Hello, did you need some help on this ?

**Comment by tobio** [comment]
@othmane399 definitely happy to get some help to get this one across the line. 

Bit of a brain dump below, but totally happy if you wanted to take another approach.

I was trying to have the provider only specify `total_shards_per_node` when it's in module definition, but that's difficult to do without dropping into the raw config objects which I'd rather avoid. 

I was going to instead validate the ES version the provider was running against and error out if the property is not supported and not the default.

**Comment by RobsonSutton** [comment]
@tobio - Apologies, have been a bit MIA recently with this one, i'll take another look at this but it may be a bit out of my comfort zone since I'm still picking up the basics currently!

File: internal/elasticsearch/index/ilm.go [hunk 1]
```diff
@@ -503,17 +509,35 @@ func expandPhase(p map[string]interface{}, d *schema.ResourceData) (*models.Phas
...
-func expandAction(a []interface{}, settings ...string) (map[string]interface{}, diag.Diagnostics) {
+var ilmActionSettingOptions = map[string]struct {
+	skipEmptyCheck bool
+	def            interface{}
+	minVersion     *version.Version
+}{
+	"number_of_replicas":    {skipEmptyCheck: true},
+	"total_shards_per_node": {skipEmptyCheck: true, def: -1, minVersion: version.Must(version.NewVersion("7.16.0"))},
+	"priority":              {skipEmptyCheck: true},
+}
```
Comment (webfella) [inline internal/elasticsearch/index/ilm.go:520]: 👍

**Review by webfella — APPROVED** [review]
LGTM 💥

**Pull request merged**